src/testDAO.mligo

```diff
@@ -0,0 +1,44 @@
+// SPDX-FileCopyrightText: 2021 TQ Tezos
+// SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+#if !TEST_H
```

Comment:
Could you please clarify why is this `testDAO` needed?

Reply:
I thought I had left a comment explaining it somewhere...

This was required to use in test where we loaded some test implementations for decision lambda. Now that we have concrete variants without lambdas in storage, this variant is one that concrete entrypoints, which just call lambdas from storage, emulating earlier functionality.

Reply:
> I thought I had left a comment explaining it somewhere...

Maybe you were thinking of this in the commit message that introduces it:
> Some tests used some predefined behavior of the lambdas. But now that
the lambdas were removed, we need a way to accomodate these tests. For
this a fourth variant, that fetched test lambdas from contract extra was
added. And in tests that checked for the core behavior, this variant was
used.

I don't see another explanation in the code.

> this variant is one that have concrete entrypoints, but just wraps calls to lambdas from storage, emulating earlier functionality for tests.

Wait, doesn't this invalidate the tests tho?
I mean the contract that is being tested isn't actually any of variants that can be compiled, right?

IMO tests that check variant-specific code should run on both `registryDAO` and `treasuryDAO` (separately), whereas the ones that don't interact with the "variable" parts can run using any of them, `trivialDAO` would be my pick.

Reply:
>Maybe you were thinking of this in the commit message that introduces it:

Yes.

>I mean the contract that is being tested isn't actually any of variants that can be compiled, right?

The registry and treasury tests uses the respective concrete contracts (not testDAO). You can see [here](https://github.com/tezos-commons/baseDAO/blob/1d5c5356181cbf0dccdbf90328cc1c05c5b833ac/haskell/test/Test/Ligo/RegistryDAO.hs#L61) for example. 

The remaining tests use testDAO, as these tests does not call any specific variant behavior, except those that test the behavior of the rest of the contract for a very specific variant code behavior. And for this lambdas loaded to testDAO is used to implement this specific behavior.

I am not sure, but it seem it is safer to use `trivialDAO` for the tests that does not need this behavior. So ll try to switch them to trivialDAO.

Reply:
I agree, `trivialDAO` seems like it would be better used for this :+1:

Reply:
Have switched to `trivialDAO` for most tests, and added a new variant `testDynamic` to implement tests that require specific behaviour of decision_callback etc, loaded as part of the test.

Reply:
I haven't checked properly yet, but couldn't these test that require a specific behavior use one of the existing callbacks?

I would imaging that the 3 variants that we have would be able to cover all these tests :thinking:

Reply:
>that require a specific behavior use one of the existing callbacks?

May be, but I think it is better to explicitly configure the behavior, than use some behavior that just incidently happen to be what we need.

Reply:
Yeah, but I mean... do we need tests on `decision callback`s behavior that **is not** the one of a supported variant?

I think all of our tests are supposed to check the behavior of one between `registry`, `treasury` and the `trivial` implementation, right?

If so, we don't need this `testDAO` IMO, because we can rely on each specific implementation that we actually want to cover. I think that this would be a better idea, as it would give us coverage of what we are actually interested in testing, rather than something that's supposed to behave in the same way.

Reply:
>do we need tests on decision callbacks behavior that is not the one of a supported variant

To be clear, we are not testing decision callback's behavior. We are testing "Given that decision callback do X,  does the rest of the contract behave in a certain way Y". So we are testing Y, not X. 

Now you are right that we can take an existing variant, and load it with some extra values, so that its decision callback will do X. I think it makes tests a little obscure. Because the intendent behavior of the callback is not as explicit, and will have to be inferred from the extra values, some proposal parameters and variant logic.  Also having the ability to pack arbitrary behavior in these callback could be useful in writing future tests.

Having said all that, I think what you suggest is not all that problematic, so I ll implement it for the sake of avoiding a drawn out discussion for a comparitively minor thing.

Reply:
Couple of problems I encounterd while trying this...

1. The decision lambda on the testDynamic contract is very liberal and does not check for the exact frozen tokens required as computed by frozen scale/extra values. It seems that without this, the 'createSampleProposal` function cannot create proposals, because it always freezes 10 tokens. 

2.  `createSampleProposal` create dummy proposals (by serializing some numbers), and when we use variants, it will try to decode them, and fail during flush. `testContract` contains a no-op decision lambda.

Let me know if you still want to proceed with this.

Reply:
> To be clear, we are not testing decision callback's behavior. We are testing "Given that decision callback do X, does the rest of the contract behave in a certain way Y". So we are testing Y, not X.

I see, but in this case, can't the straightforward implementation of `trivialDAO` be used for that?

AFAICT all it does is return the input `extra`, so it should be possible to:
- test the "Y" without being concerned by what the decision lambda does
- keep creating dummy proposals in `createSampleProposal` (I assume it has to be changed a bit, but it should still be possible to generate these quite randomly).
- predict the frozen tokens needed to be used, even if that's a bit more annoying than allowing any value (it probably a good thing to not be so liberal here too)

> Let me know if you still want to proceed with this.

Yep, I think we should.
Sorry for dragging this out so much, but IMO these tests were maybe only a little odd before, but quite a noticeable exception now.

I know this PR has been going for a long time, but since it's close to being completed now I think that it would be good to polish these tests while we are at it and avoid pushing this `testDAO` to `master` to begin with.

Reply:
No problem. I have removed the test variant and replaced its use with TrivialDAO. I had to remove a couple of tests, which does not appear to test anything significant.